### PR TITLE
Form pillow to mark has_submissions

### PR DESCRIPTION
**Pull request by benrudolph**
@czue here's a quick WIP for the pillow to mark apps as having submissions. not sure if it's worth a whole new pillow but couldn't immediately find a good pillow to put it in. also had a really hard time naming some of these classes

### CODE REVIEW

File: corehq/pillows/xform.py [hunk 1]
```diff
@@ -163,6 +164,22 @@ def get_couch_form_reindexer():
...
 
+def get_form_kafka_pillow(pillow_id='form-processor-pillow'):
+    checkpoint = PillowCheckpoint(
+        pillow_id,
+    )
+    form_processor = FormProcessor()
+    return ConstructedPillow(
+        name=pillow_id,
+        checkpoint=checkpoint,
+        change_feed=KafkaChangeFeed(topics=[topics.FORM], group_id='domains-to-es'),  # TODO: include multiple topics
```
Comment (benrudolph) [inline corehq/pillows/xform.py]: do you think this'll be hard to implement?

Comment (dimagimon) [inline corehq/pillows/xform.py]: E501 line too long (117 > 115 characters)

Comment (benrudolph) [inline corehq/pillows/xform.py]: (also wrong group_id)

File: corehq/ex-submodules/pillowtop/processors/form.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        if change.deleted and change.id:
+            return
+        doc = change.get_document()
+        if not doc:
+            return
+
+        build_id = doc.get('build_id')
+        domain = change.get_document().get('domain')
+
+        if not build_id or not domain:
```
Comment (czue) [inline corehq/ex-submodules/pillowtop/processors/form.py]: should this not be `if build_id and domain`?

File: corehq/ex-submodules/pillowtop/processors/form.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+from corehq.util.quickcache import quickcache
+
+
+class FormProcessor(PillowProcessor):
+    """
+    Pillow used to process each form and make updates if necessary
+    """
+
+    def process_change(self, pillow_instance, change):
+        if change.deleted and change.id:
```
Comment (czue) [inline corehq/ex-submodules/pillowtop/processors/form.py]: why `change.id`?

File: corehq/ex-submodules/pillowtop/processors/form.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            return
+        doc = change.get_document()
+        if not doc:
+            return
+
+        build_id = doc.get('build_id')
+        domain = change.get_document().get('domain')
```
Comment (czue) [inline corehq/ex-submodules/pillowtop/processors/form.py]: canonical way to get the domain would be `change.metadata.domain`

File: corehq/ex-submodules/pillowtop/processors/form.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+        if not build_id or not domain:
+            app = self._get_app(domain, build_id)
+            app.has_submissions = True
+            app.save()
```
Comment (czue) [inline corehq/ex-submodules/pillowtop/processors/form.py]: this should probably check whether `has_submissions` is already set and not save the doc if it is.

File: corehq/pillows/xform.py (hunk 1 above)
Comment (czue) [inline corehq/pillows/xform.py]: will what be hard?

**Comment by czue** [comment]
good question on the new pillow. I think it's probably ok/the right architecture for now though if we get too many pillows our monoliths are going to get unhappy. agree the names are tough (sorry lots of awkward generic component classes). anyway, this all looks pretty straightforward to implement once the details are ironed out. will probably leave it for when you come back unless icds can't deal with the workaround

File: corehq/ex-submodules/pillowtop/processors/form.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            return
+
+    @quickcache(['domain', 'build_id'], timeout=60 * 60)
+    def _get_app(self, domain, build_id):
+        return get_app(domain, build_id)
```
Comment (snopoke) [inline corehq/ex-submodules/pillowtop/processors/form.py]: Would need to clear the cache after save

File: corehq/pillows/xform.py (hunk 1 above)
Comment (benrudolph) [inline corehq/pillows/xform.py]: i want to include multiple topics (both sql and couch forms) and i don't think that's implemented yet

File: corehq/ex-submodules/pillowtop/processors/form.py (hunk 2 above)
Comment (benrudolph) [inline corehq/ex-submodules/pillowtop/processors/form.py]: 😳 will write tests in final version haha

**Comment by benrudolph** [comment]
thanks for the feedback. didn't actually think the base class names were that hard to grok; i was more commiserating about the fact that I named a pillow `FormPillow` which is extremely generic for a task that's very specific

File: corehq/ex-submodules/pillowtop/processors/form.py (hunk 6 above)
Comment (benrudolph) [inline corehq/ex-submodules/pillowtop/processors/form.py]: 👍 makes sense

File: corehq/pillows/xform.py (hunk 1 above)
Comment (snopoke) [inline corehq/pillows/xform.py]: Handling multiple topics is easy. See ledger pillow as an example.

Not sure if we're publishing couch forms to kafka yet though. But that isn't hard either.  See DomainDbKafkaPillow.

Comment (czue) [inline corehq/pillows/xform.py]: the main db publishes to kafka so all cases and forms are there.

**Comment by czue** [comment]
re naming, you could just name it `AppFormSubmissionTrackerPillow` or something with a comment that it could be used for something more generic down the road. one question i don't know the answer to is whether if we had to do another thing with forms whether we would lump that in the same pillow or whether we'd make a new pillow. i think would be a tradeoff between responsibility isolation and system resources/performance.

**Comment by snopoke** [comment]
I've also been thinking about that. If we did lump them together then we'd
want to have a structure that would allow us to isolate the different
pieces of functionality (like different processor classes that could be
registered to the same pillow). It would also be cool if we could make use
of gevent so that the processors can run concurrently.

**Comment by benrudolph** [comment]
i'm in the lumping together camp since we could theoretically build an abstraction architecture that would separate responsibilities as simon suggests. running many pillows seems non ideal, in my head they feel fairly heavyweight. it already takes ages to shut down the pillow processes

**Comment by benrudolph** [comment]
Ok i think this is ready for re-review

**Comment by czue** [comment]
for what it's worth I'm less clear which camp I'm in. stuff like checkpointing, retrying, and reindexing could get pretty messy if we lump everything together into megapillows (funny mental image), and we already have the abstraction layer of the topic within kafka which provides a nice clean separation and mechanisms for a lot of that stuff. 

in my head what would be most ideal is if we could get the overhead of running a bunch of pillows down. potentially by lumping them together in a singe python/gevent process but still keeping the pillow isolation in tact, or (in an unlikely dream scenario) getting the overhead of running a single one way down.

**Comment by czue** [comment]
build fail real (you just need to update the test).
```
======================================================================
FAIL: test_instantiate_all (testapps.test_pillowtop.tests.test_settings.PillowtopSettingsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/testapps/test_pillowtop/tests/test_settings.py", line 35, in test_instantiate_all
    self.assertEqual(len(all_pillow_configs), len(expected_meta))
AssertionError: 49 != 48
----------------------------------------------------------------------
```

**Comment by benrudolph** [comment]
good point, i think in my head i immediately assumed that it'd be very difficult to make the pillow processes lightweight. i think it'd be a huge gain for us as a whole to reduce the overhead of starting and stopping pillows

**Comment by benrudolph** [comment]
at the time of writing this i've been waiting 10 minutes for our pillows to start 😓

**Comment by snopoke** [comment]
That could also be related to this:
https://github.com/dimagi/commcare-hq/pull/12030

If the kafka log is big then getting the highwater mark can take a long
time since we go through the entire log. I haven't figured out why my
change causes the failure it does yet.

File: corehq/ex-submodules/pillowtop/processors/form.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        if build_id and domain:
+            app = self._get_app(domain, build_id)
+
+            if not app.has_submissions:
+                app.has_submissions = True
+                app.save()
+                self._get_app.clear(self, domain, build_id)
+
+    @quickcache(['domain', 'build_id'], timeout=60 * 60)
+    def _get_app(self, domain, build_id):
```
Comment (czue) [inline corehq/ex-submodules/pillowtop/processors/form.py]: can you not just cache the `has_submissions` flag? seems like that would be a lot more memory efficient (and you'd only have to get the app on a cache miss, or when you wanted to save it)

File: testapps/test_pillowtop/tests/test_form_pillow.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+from corehq.apps.change_feed.consumer.feed import change_meta_from_kafka_message
+from corehq.apps.change_feed.tests.utils import get_test_kafka_consumer, get_current_kafka_seq
+from corehq.apps.receiverwrapper import submit_form_locally
+from corehq.apps.userreports.tests.utils import doc_to_change
+from corehq.pillows.xform import get_app_form_submission_tracker_pillow
+from corehq.form_processor.tests.utils import FormProcessorTestUtils
+from corehq.form_processor.utils import TestFormMetadata, get_simple_form_xml
+
+
+class FormPillowTest(TestCase):
```
Comment (czue) [inline testapps/test_pillowtop/tests/test_form_pillow.py]: nice tests!

File: testapps/test_pillowtop/tests/test_form_pillow.py [hunk 9]
```diff
@@ -90,6 +90,23 @@ def test_form_pillow_mismatch_domains(self):
...
+
+        # confirm change made it to kafka
+        self.assertFalse(self.app.has_submissions)
+
+        self.pillow.process_changes(since=kafka_seq, forever=False)
+        self.assertTrue(Application.get(self.app._id).has_submissions)
+
+        self._make_form()
+        self.pillow.process_changes(since=kafka_seq, forever=False)
+        self.assertTrue(Application.get(self.app._id).has_submissions)
```
Comment (czue) [inline testapps/test_pillowtop/tests/test_form_pillow.py]: could also check that the `_rev` hasn't changed to ensure we're not doing duplicate saves

**Comment by czue** [comment]
no blockers but will give you a chance to respond to the cache suggestion before merge. otherwise :+1: on :white_check_mark: from me

File: testapps/test_pillowtop/tests/test_form_pillow.py (hunk 9 above)
Comment (benrudolph) [inline testapps/test_pillowtop/tests/test_form_pillow.py]: good call, will update

File: corehq/ex-submodules/pillowtop/processors/form.py (hunk 7 above)
Comment (benrudolph) [inline corehq/ex-submodules/pillowtop/processors/form.py]: ahhh that's a good idea

Comment (benrudolph) [inline corehq/ex-submodules/pillowtop/processors/form.py]: https://github.com/dimagi/commcare-hq/pull/11819/commits/dbe1b914d98fc80cf6d6cf85831502e1ea93ccb0
is this what you were thinking? this way we only grab the app if it hasn't already been marked

File: testapps/test_pillowtop/tests/test_form_pillow.py (hunk 9 above)
Comment (benrudolph) [inline testapps/test_pillowtop/tests/test_form_pillow.py]: https://github.com/dimagi/commcare-hq/pull/11819/commits/1a01cf1d318bddc40512124f34d5b67e70cd9fe3

**Comment by benrudolph** [comment]
✅

File: corehq/ex-submodules/pillowtop/processors/form.py [hunk 10]
```diff
@@ -24,16 +24,16 @@ def process_change(self, pillow_instance, change):
...
-
-            if not app.has_submissions:
-                app.has_submissions = True
-                app.save()
-                self._get_app.clear(self, domain, build_id)
+            self._mark_has_submission(domain, build_id)
 
     @quickcache(['domain', 'build_id'], timeout=60 * 60)
-    def _get_app(self, domain, build_id):
+    def _mark_has_submission(self, domain, build_id):
```
Comment (czue) [inline corehq/ex-submodules/pillowtop/processors/form.py]: clever and unintuitive use of caching. could maybe add a comment explaining it.

**Pull request merged**